### ISSUE TITLE
Images

### ISSUE THREAD

**Issue by whitehat101** [gorn/rspreadsheet#24]
For the last part of my project, I needed to insert an image. Adding an image requires modifying several files inside the ods archive, more than just `content.xml`. I found `randym/axlsx` supports image insertion, but I wasn't able to make sense of enough of their code (their project is builder-only).

In the end, I made a template ods with a placeholder image of fixed format, dimensions, etc. Then I monkey patched a new save method to overwrite the image while when saving the zip.

```ruby
# definitely a hack, but it worked for me
    def initialize
      @book = Rspreadsheet.open 'template/path.ods'
      @sheet = @book.worksheets 1

      def @book.save_with_image fname, b64image
        FileUtils.cp @filename, fname
        Zip::File.open(fname) do |zip|
          # it is easy, because @xmlnode in in sync with contents all the time
          zip.get_output_stream('content.xml') do |f|
            f.write @content_xml.to_s(:indent => false)
          end
          if b64image
            index = b64image.index 'base64,'
            b64image = b64image[(index+7)..-1] if index
            zip.get_output_stream('Pictures/1000020100000524000000FF5BDFEF915A06D909.png') do |f|
              f.write Base64.decode64(b64image)
            end
          end
        end
      end
```

The output_stream filename was manually discovered by inspection of `content.xml`, or by unpacking the ods archive.

If image support is as complex as I suspect it might be, it might be nice to have an API to manually (over)write files into the archive, to band-aid this feature.

**Comment by gorn** [comment]
Thanks for sharing this. Although definitly a hack it might serve as a steppingstone for the general feature. Just curious: what exactly is the b64image? Its a string you generate in your code, or a file read from disk? Wjat does the "path.ods" file contains (maybe attach a sample here).

**Comment by whitehat101** [comment]
Ah, that was a bit my-app specific. It's a [Data URI](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/Data_URIs). Mine was generated by a HTML `<canvas>` and `canvas.toDataURL()`.

For a general API, you'd probably want to accept stuff that acts like an `IO`, and not do decoding like that inside the Zip block.

```ruby
def save(new_filename_or_io_object=nil, files = {})
  # ...
  Zip::File.open(@filename) do |zip|
    zip.get_output_stream('content.xml') do |f|
      f.write @content_xml.to_s(:indent => false)
    end
    files.each do |name, io|
      zip.get_output_stream(name) do |f|
        f.write io.read
      end
    end
  end
end

# use
save 'new.ods', "Pictures/..." => io
```

[examples.zip](https://github.com/gorn/rspreadsheet/files/647372/examples.zip)
- a blank spreadsheet
- an image starting in B2 (anchored to page)
- an image starting in B2 (anchored to cell)
- a similar png to the embedded image

**Comment by gorn** [comment]
@whitehat101 I had implemented very basic images handling. Are you willing to test it a bit?

Worksheet has now methods 
```
images
images(index)
insert_image_to(x,y,filename)
```
and some more. A code is bit ugly, but I would like to write some more tests before I start to clean up and release it. Currently you must pull it from github.

**Comment by gorn** [comment]
Images are implemented in [release 0.3](https://github.com/gorn/rspreadsheet/releases/tag/v0.3)
